File: leetcode/medium/1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.js
// https://leetcode.com/problems/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/
// Definition for a binary tree node.
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}
/**
 * @param {TreeNode} original
 * @param {TreeNode} cloned
 * @param {TreeNode} target
 * @return {TreeNode}
 */

function getTargetCopy(original, cloned, target) {
    return dfs(cloned, target);
};

function dfs(cloned, target) {
    if (cloned) {
        if (cloned.val === target.val) return cloned;
        return dfs(cloned.left, target) || dfs(cloned.right, target);
    }
    return null;
}

function bfs(original, cloned, target) {
    let q = [[original, cloned]];

    while (q.length) {
        const next = [];
        for (let [o, c] of q) {
            if (o.val === target.val) return c;
            if (o.left) next.push([o.left, c.left]);
            if (o.right) next.push([o.right, c.right]);
        }
        q = next;
    }
}

File: leetcode/medium/1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.cc
// https://leetcode.com/problems/find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/
#include <iostream>

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        if (original == NULL) return NULL;
        if (original == target) return cloned;

        TreeNode* cp = getTargetCopy(original->left, cloned->left, target);
        if (cp != NULL && cp->val == target->val) return cp;

        TreeNode* cp2 = getTargetCopy(original->right, cloned->right, target);
        return cp2;
    }
};

int main() {
    return 0;
}